tempesta_fw/ss_skb.c

```diff
@@ -97,6 +98,28 @@ ss_skb_alloc_pages(size_t len)
 	return skb;
 }
 
+/**
+ * Given the total message length as @len, allocate an appropriate number
+ * of SKBs and page fragments to hold the payload, and add them to the
+ * message. Put as much as possible in one SKB. TCP GSO will take care of
+ * segmentation. The allocated payload space will be filled with data.
+ */
+int
+ss_skb_alloc_data(struct sk_buff **skb_head, size_t len)
+{
+	int i_skb, nr_skbs = DIV_ROUND_UP(len, SS_SKB_MAX_DATA_LEN);
+	struct sk_buff *skb;
+
+	for (i_skb = 0; i_skb < nr_skbs; ++i_skb) {
+		skb = ss_skb_alloc_pages(min(len, SS_SKB_MAX_DATA_LEN));
```

Comment:
The variable `len` does not decrease, so it seems that the last skb always will be allocated with `MAX_SKB_FRAGS` quantity of frags, even if remaining data will fits entirely in the linear part of skb.